### PR TITLE
[Abuse CH] Add Threat Fox Datastream

### LINKED ISSUE #3949: Abuse.ch | Threat Fox
### Description
Our current Abuse.ch provides coverage for Threat Intel data from URL based indicators, MalwareBazaar and more. Abuse.ch recently released Threat Fox, a community driven project where security researchers and threat analysts can share indicators of compromise (IOCs) with the infosec community. Domains, IP addresses, email addresses and file hashes associated with malware, botnet command&control (C&C), payload or payload delivery can all be shared via ThreatFox. 

### Architecture
ThreatFox has an API we can use to ingest IoC data: https://threatfox.abuse.ch/api/

# Integration release checklist

This checklist is intended for integrations maintainers to ensure consistency
when creating or updating a Package, Module or Dataset for an Integration.

### All changes

- [ ] Change follows the [contributing guidelines](https://github.com/elastic/integrations/blob/master/CONTRIBUTING.md)
- [ ] Supported versions of the monitoring target are documented
- [ ] Supported operating systems are documented (if applicable)
- [ ] Integration or [System tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/system_testing.md) exist
- [ ] Documentation exists
- [ ] Fields follow [ECS](https://github.com/elastic/ecs) and [naming conventions](https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html)
- [ ] At least a manual test with ES / Kibana / Agent has been performed.
- [ ] Required Kibana version set to:

### New Package

- [ ] Screenshot of the "Add Integration" page on Fleet added

### Dashboards changes

- [ ] Dashboards exists
- [ ] Screenshots added or updated
- [ ] Datastream filters added to visualizations

### Log dataset changes

- [ ] [Pipeline tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/pipeline_testing.md) exist (if applicable)
- [ ] Generated output for at least 1 log file exists
- [ ] Sample event (`sample_event.json`) exists

### PR DESCRIPTION
**Pull request by legoguy1000**
## What does this PR do?

Add Threat Fox Datastream to Abuse CH Integration

## Checklist

- [X] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [X] I have verified that all data streams collect metrics or logs.
- [X] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ]

## Related issues

- closes #3949

### CODE REVIEW

File: packages/ti_abusech/data_stream/threatfox/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+          confidence = "Medium";
+        } else if (value >= 70 && value <= 100) {
+          confidence = "High";
+        }
+        ctx.threat.indicator.put("confidence", confidence)
+  ## URL/URI indicator operations
+  - set:
+      field: threat.indicator.type
+      value: url
+      if: "ctx.abusech?.threatfox?.ioc_type != null && ['url'].contains(ctx.abusech?.threatfox?.ioc_type)"
```
Comment (efd6) [inline packages/ti_abusech/data_stream/threatfox/elasticsearch/ingest_pipeline/default.yml:110]:
```suggestion
      if: "ctx.abusech?.threatfox?.ioc_type != null && ['url'].contains(ctx.abusech.threatfox.ioc_type)"
```

File: packages/ti_abusech/data_stream/threatfox/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+      target_field: threat.indicator.url
+      keep_original: true
+      remove_if_successful: true
+      if: ctx.abusech?.threatfox?.ioc_type == 'url'
+
+  ## IP/Port indicator operations
+  - set:
+      field: threat.indicator.type
+      value: domain-name
+      if: "ctx.abusech?.threatfox?.ioc_type != null && ['domain'].contains(ctx.abusech?.threatfox?.ioc_type)"
```
Comment (efd6) [inline packages/ti_abusech/data_stream/threatfox/elasticsearch/ingest_pipeline/default.yml:122]:
```suggestion
      if: "ctx.abusech?.threatfox?.ioc_type != null && ['domain'].contains(ctx.abusech.threatfox.ioc_type)"
```

File: packages/ti_abusech/data_stream/threatfox/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  - rename:
+      field: abusech.threatfox.ioc
+      target_field: threat.indicator.url.domain
+      ignore_missing: true
+      if: "ctx.abusech?.threatfox?.ioc_type != null && ['domain'].contains(ctx.abusech?.threatfox?.ioc_type)"
```
Comment (efd6) [inline packages/ti_abusech/data_stream/threatfox/elasticsearch/ingest_pipeline/default.yml:127]:
```suggestion
      if: "ctx.abusech?.threatfox?.ioc_type != null && ['domain'].contains(ctx.abusech.threatfox.ioc_type)"
```

File: packages/ti_abusech/data_stream/threatfox/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  - set:
+      field: threat.indicator.type
+      value: ipv4-addr
+      if: "ctx.abusech?.threatfox?.ioc_type != null && ['ip:port'].contains(ctx.abusech?.threatfox?.ioc_type)"
```
Comment (efd6) [inline packages/ti_abusech/data_stream/threatfox/elasticsearch/ingest_pipeline/default.yml:131]:
```suggestion
      if: "ctx.abusech?.threatfox?.ioc_type != null && ['ip:port'].contains(ctx.abusech.threatfox.ioc_type)"
```

File: packages/ti_abusech/data_stream/threatfox/fields/fields.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  description: All fields related to AbuseCH Threat Fox indicators.
+  fields:
+    - name: tags
+      type: keyword
+      description: >
+        A list of tags associated with the queried malware sample.
+
+    - name: confidence_level
+      type: long
+      description: "Confidence level between 0-100      \n"
```
Comment (efd6) [inline packages/ti_abusech/data_stream/threatfox/fields/fields.yml:12]: Is this intended?

**Comment by legoguy1000** [comment]
looks good to merge.

**Review by andrewkroh — APPROVED** [review]

**Pull request merged**